core/src/main/java/com/predic8/membrane/core/transport/http/MessageTracer.java

```diff
@@ -0,0 +1,98 @@
+package com.predic8.membrane.core.transport.http;
+
+import com.predic8.membrane.core.http.*;
+import org.slf4j.*;
+
+/**
+ * Writes Request and Response messages to the trace log.
+ *
+ * Attention: Possible data leakage!
+ * Tracer may write sensitive values to the log like Authorization headers and body content.
+ *
+ * Keep this class at that location to keep the same package for logger configuration!
+ */
+public class MessageTracer {
+
+    private static final Logger log = LoggerFactory.getLogger(MessageTracer.class);
+
+    private static final int TRACE_BODY_MAX_LEN = getTraceBodyMaximumLength();
+    public static final String MEMBRANE_MESSAGE_TRACER_MAX_BODY_LENGTH = "membrane.message.tracer.maxBodyLength";
+
+    /**
+     * Writes Request and Response messages to the trace log
+     */
+    public static void trace(Message message) {
+        if (!log.isTraceEnabled())
+            return;
+
+        if (message instanceof Request request)
+            traceRequest(request);
+        if (message instanceof Response response)
+            traceResponse(response);
+    }
+
+    private static void traceRequest(Request req) {
+        try {
+            StringBuilder sb = new StringBuilder("\n---[ HTTP Request ]----------------------------\n");
+            sb.append(req.getStartLine())                         // e.g. "GET /foo HTTP/1.1"
+                    .append('\n')
+                    .append(req.getHeader().toString());                // all header fields
+
+            // append body if we actually have one and it is already buffered
+            if (!req.isBodyEmpty()) {
+                sb.append('\n');
+                appendBody(sb, req.getBodyAsStringDecoded());
```

Comment:
This could be optimized: `getBodyAsStringDecoded()` loads the complete body into a String (which is not optimal), but only the first `TRACE_BODY_MAX_LEN` chars are output anyways.